src/ocrd/processor/base.py

```diff
@@ -323,24 +325,31 @@ def process_page_file(self, *input_files : OcrdFile) -> None:
         to handle cases like multiple fileGrps, non-PAGE input etc.)
         """
         log = getLogger('ocrd.processor.base')
-        input_pcgts : List[OcrdPage] = [None] * len(input_files)
+        input_pcgts : List[Optional[OcrdPage]] = [None] * len(input_files)
+        assert isinstance(input_files[0], (OcrdFile, ClientSideOcrdFile))
```

Comment:
Why `Optional` (also in function prototype)?

Reply:
Here: Because we're instantiating a list of `None` values, which are not `OcrdPage`.

In the function signature of `process_page_pcgts`: Same situation, there might be "holes" in the list of `input_pcgts` when any of the `input_files` in `process_page_files` cannot be parsed as PAGE-XML.

And for `process_page_files`: The `input_files` can be hole-y, if the `workspace.download_file` fails for any of the files (beyond the first?).

But really, I was trying to make sure that static type checking had no more complaints. I tried to add `assert` statements where I know that variables must be defined or of a certain type to mitigate the "everything might be `None`" problem somewhat.

Reply:
Oh, right, I forgot about the holes returned by zip_input_files for multiple fileGrps but incomplete PAGE-XML coverage per page!

Maybe we should document this more loudly.